### ISSUE TITLE
Updating search params causes stale values / infinite re-renders

### ISSUE THREAD

**Issue by Moshyfawn** [TanStack/router#453]
### Describe the bug

Using a controlled input with URL search params yields infinite re-renders / stale search params value when updating via `navigate({ search: { param: newValue } })`

### Your Example Website or App

https://codesandbox.io/p/sandbox/router-vite-4-build-preview-404-forked-o7q8tu?selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A9%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A9%7D%5D&file=%2Fsrc%2FApp.tsx

### Steps to Reproduce the Bug or Issue

1. Go to [repro](https://codesandbox.io/p/sandbox/router-vite-4-build-preview-404-forked-o7q8tu?selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A9%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A9%7D%5D&file=%2Fsrc%2FApp.tsx)
2. Enter a single digit number into the "floor" input
3. See error: search param "floor" in the URL has changed to the value entered, but input value remains empty _(or prev value if had already entered a value previously)

P.S. In my app locally, the behaviour actually causes infinite re-renders; In sandbox I was only able to repro the "stale value" behaviour

### Expected behavior

I expect the search param state to update after calling `navigate` with new search param value, so that the value and the search param in the URL are in sync; the behaviour doesn't cause infinite re-renders.

Including the screen-cast of the behaviour below as well

### Screenshots or Videos

https://user-images.githubusercontent.com/16290753/207154938-0b9ca848-b774-448d-a9c8-1495cac5f88f.mp4

### Platform

- OS: macOS
- Browser: Firefox
- Version: 108.0b9

### Additional context

```json
"dependencies": {
  "@tanstack/react-router": "0.0.1-beta.38",
  "react": "^18.2.0",
  "react-dom": "^18.2.0",
  "zod": "^3.20.1"
},
```

**Comment by ogchefi** [comment]
Also experiencing stale values. Does not occur every time, for example this counter works only now and then:

```
const useGameCounter = () => {
  const [timesUp, setTimesUp] = useState(false)
  const { search, navigate } = useMatch('/play')

  useEffect(() => {
    console.log('search.counter:', search.counter)
    if (search.counter > 0) {
      console.log('set timeout')
      const timeout = setTimeout(() => {
        console.log('navigate')
        navigate({
          search: (old) => {
            return {
              ...old,
              counter: old.counter - 1,
            }
          },
        })
      }, 1000)
      return () => {
        console.log('unmount')
        clearTimeout(timeout)
      }
    } else {
      setTimesUp(true)
    }
  }, [navigate, search.counter])

  return { timesUp }
}
```

In console it clearly calls navigate last and reduces the counter in url, but the change does not get reflected to React.

**Comment by ogchefi** [comment]
The counter actually updates every time I focus out of the Chrome browser window and then back. It's possible to walk down the counter this way.

**Comment by judicaelandria** [comment]
can you update and check again if the error still occurs please?

**Comment by Moshyfawn** [comment]
Oh, the API has drastically changed. Give me a moment to learn the new way to build routes

**Comment by Moshyfawn** [comment]
I've [forked the sandbox](https://codesandbox.io/p/sandbox/router-search-default-forked-3mveqk?file=%2Fsrc%2FApp.tsx&selection=%5B%7B%22endColumn%22%3A13%2C%22endLineNumber%22%3A8%2C%22startColumn%22%3A13%2C%22startLineNumber%22%3A8%7D%5D) and updated the fork to match the current API but I'm yet to run it and check if the problem still persists as I can't make sandbox project to re-compile, for some reason

**Comment by Moshyfawn** [comment]
Alright, it looks like it's fixed in the latest (`v0.0.1-beta.61`) package version.

[Here's a working sandbox](https://codesandbox.io/p/sandbox/router-search-default-forked-3mveqk?file=%2Fpackage.json&selection=%5B%7B%22endColumn%22%3A46%2C%22endLineNumber%22%3A12%2C%22startColumn%22%3A33%2C%22startLineNumber%22%3A12%7D%5D)

_P.S. The sandbox editor + preview view is bugged for me, but opening the demo app in the separate window works_
